designs/odcr.md

```diff
@@ -0,0 +1,408 @@
+
+## Pricing/Consolidation
+
+### Provisioning
+
+Pricing is directly considered during provisioning and consolidation as capacity-reservation is prepaid. It is assumed to have a price of "near 0" during provisioning (see [Consolidating between Capacity Reservations](#consolidating-between-capacity-reservations) for more detail on why this is not just "0").
+
+### Consolidation
+
+During consolidation pricing does matter as it affects which candidate will be [prioritized](https://github.com/kubernetes-sigs/karpenter/blob/75826eb51589e546fffb594bfefa91f3850e6c82/pkg/controllers/disruption/consolidation.go#L156). Since all capacity instances are paid ahead of time, their cost is already incurred. Users would likely want to prioritize filling their reserved capacity.
+reservation first then fall back into other instances. Because of this reserved instances should likely show up as 0 dollar pricing when we calculate the instance pricing.
+
+#### Consolidating into Capacity Reserved Instances
+
+If we track Capacity Reservation usage, we can optimize the cluster configuration by moving non-Capacity Reserved instances into
+Capacity Reserved instances. We would need to match the instance type, platform and availability zone prior to doing this.
+
+This would be done by the standard consolidation algorithm and should work with minimal changes, since consolidation already optimizes for cost. 
+
+#### Consolidating between Capacity Reservations
+
+Treating the price of all capacity reservation offerings as `0` sounds sensical but has some adverse interactions with consolidation. Most notably, if we launch an instance into a capacity reservation offering, we will _never_ consolidate out of that offering -- even if there is a smaller instance type that would work just as well for the pods on that node.
+
+In practicality, that larger capacity reservation could have been used for other work -- perhaps on another cluster, but may have been held by a single pod that prevented Karpenter from consolidating it.
+
+To solve for this edge case, we won't model the pricing of capacity reservations as `0` but as a "near-0" value. We'll divide the existing price of the on-demand offering by 10,000,000 to represent a price that is less than every other instance type offering, but still maintains the relative ordering of on-demand instance types.
+
+_NOTE: To contrive the number we use here, we can take the most expensive hourly rate for an on-demand instance ($407.68 for the `u7in-32tb.224xlarge`) and the least expensive hourly rate for a spot instnace ($0.0015 for `t4g.nano`) to calculate `407.68/0.0015 = 271,787`). We then add some buffer to this number with a couple orders of magnitude to get to `10,000,000`_
```

Comment:
but if there's a 2240xlarge sometime in the future, we may need to reconsider!!!

Reply:
another weird thing is that if someone was taking an ODCR on a huge instance, but only had a super tiny pod, it costs less from the cluster perspective, but more expensive from the overall perspective. Nothing to do here about that though.

Reply:
Yeah, eventually other pods would probably schedule -- agree that there's not a ton that we could do here and what we have is a good starting point.

Reply:
Good call -- agreed that we can't make this a static value and more than likely have to take our current pricing data and just add some orders of magnitude to it